agent/doctor/healthcheck.go

```diff
@@ -0,0 +1,70 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//      http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package doctor
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi"
+	"github.com/cihub/seelog"
+)
+
+type Healthcheck interface {
+	GetHealthcheckStatus() HealthcheckStatus
+	GetHealthcheckTime() time.Time
+	RunCheck() bool
+}
+
+type dockerRuntimeHealthcheck struct {
+	// Status is the container health status
+	LatestCheckStatus HealthcheckStatus `json:"latestStatus,omitempty"`
+	// Since is the timestamp when container health status changed
+	LatestCheckTime time.Time `json:"latestStatusTime,omitempty"`
+	// Status is the container health status
+	LastCheckStatus HealthcheckStatus `json:"lastStatus,omitempty"`
+	// Since is the timestamp when container health status changed
+	LastCheckTime time.Time `json:"lastStatusTime,omitempty"`
+	client        dockerapi.DockerClient
+	lock          sync.RWMutex
+}
+
+func NewDockerRuntimeHealthcheck(client dockerapi.DockerClient) *dockerRuntimeHealthcheck {
+	return &dockerRuntimeHealthcheck{
+		client:            client,
+		LatestCheckStatus: HealthcheckStatusInitializing,
```

Comment:
If these are meant to only be read by the public and read-locked functions (GetHealthcheckStatus), then these should not be public.

Reply:
these follow the pattern in our api code and allow us to use the json struct field tags.
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/container/container.go#L102-L111

Reply:
why would we follow the pattern in our API code? this code is more internal isn't it?